### ISSUE TITLE
Spike: API design to support initial group-selector improvements

### ISSUE THREAD

**Issue by lyzadanger** [hypothesis/product-backlog#901]
Part of https://github.com/hypothesis/product-backlog/issues/865

For the first phase of the Community Groups work, we're allowing annotation within certain configured restricted groups outside of their defined scopes. There are some changes to the client's group-selector menu that will be going along with that.

This issue is concerned with the extension/design of our API to satisfy those needs.

I attach the following **extremely WIP FPO example only** design notion for what this might look like:

![menucomp](https://user-images.githubusercontent.com/439947/50299796-81051700-0450-11e9-9609-7778cc0251d2.png)

In the above comp, "Featured Groups" is a list of groups that is _variable_: the user might see different featured groups depending on where they are on the Web. These groups include scoped open groups, as well as scoped restricted groups to which the user is _not_ a member.

"My Groups" is a list of groups that should not vary; this list stays constant web-wide and is a representation of all of the groups that the user _is a member of_ as well as the "special" Public group for first party users—it can be argued that, from a user's point of view, they _are_ a member of Public, even if technically that is not the case.

So: the "My Groups" list contents do not change, but there is a wrinkle that needs to be accounted for. We don't want the "My Groups" list to be switching around confusingly across domains, but there will be times when certain restricted groups will not be writeable because of scope restrictions. The above comp shows a proposed solution for this, visually, but implies a few additional needs from our API response(s) that we need to accommodate.

### How this is different from now

At present, restricted groups are only returned within their defined scopes, no matter what the user's membership relationships are with those groups. They appear and disappear in a member's group menu just as they do for non-members. We need to make it so that members see all of their groups, no matter where they are on the web (even if they can't put annotations in all of those groups)—thus, the overall list of groups returned to the client for display in the group-selector menu needs to change.

The current Group API resource includes properties that indicate the group's `type` (restricted, open, private) and a simple Boolean `scoped` property that is an indicator that the group has one or more scopes (not abundantly useful TBH). More details on the group resource can be had by going [to the docs](https://h.readthedocs.io/en/latest/api-reference/#tag/groups/paths/~1groups~1{id}/get)  and expanding the "Response" section.

 In the new requirements, however, the client will need some additional information for each group:

* Scopes: Not merely that the group is scoped or not, but what those scopes are
* Whether the current user may add annotations to the group based on that user's membership in the group and the current document URI (i.e. Can _this_ user add annotations to _this_ group at _this_ URL?)

### The Questions the API is Answering Now

Currently, our `GET /api/groups` endpoint answers several questions at once, somewhat implicitly. It is tasked with retrieving:

* All of the open and restricted groups currently in scope based on the document URI
* The Public group (in 1st-party contexts)
* All of the user's Private groups

In addition, it has a bunch of client-specific logic regarding the ordering of those groups. Thus, it answers the question "What groups should this user see in this document URI context and in what order in the `h` client?" but doesn't provide any information as to _why_ or the user's current relationship to a given group. There is no, ah, grouping or taxonomy of the groups; they're just returned in a flat Array in the order that the service has deemed appropriate for _our_ client.

### The Natural Questions I think the API Could Answer...

Instead, I find it helpful to think about the list of groups to be the answer of two separate questions—one about context/document/scope and one about membership:

1. What groups are relevant to the current context (i.e. _document URI_)? This corresponds to  "Featured Groups"/scoped groups. The current user's membership in these groups is not relevant.
1.  What groups are relevant to this _user_? What is the list of groups of which this user is a member, plus the special "Public" group—because from the user's perspective, they _are_ in effect a member of "Public".

(In case you've observed that one given group could appear as part of both answers—you're right! A group may be both a "featured group"—in scope—and one of a user's groups by membership).

## TL;DR

Point being, we need to figure out an API design that both satisfies the new UX needs as well as provides a reasonable modeling of resources.

**Comment by lyzadanger** [comment]
All right! I think we have a proposed API design!

Changes needed:

* `group` resource: make it possible to expand `scopes` information the way that you can expand `organization` information in API responses. Be careful not to bake too much about our current scopes implementation into resources.
* Add new endpoint (route/path TBD) to retrieve all groups to which the currently-auth'd user belongs (i.e. the "My Groups" list). I'm going with the approach right now that if there is an auth'd user, this response should include the `Public` group for first-party users as, from a user's perspective, they _are_ a member of that group, and it should show up in the "My Groups" list. We can debate this detail as we go. This is a useful endpoint to have in general.
* No changes in this phase to the existing `GET /api/groups` endpoint.
* To support community-groups features, the client will make two concurrent requests to the `GET /api/groups` and new user-groups endpoint. It will concoct the "Featured Groups" and "My Groups" lists by traversing both API responses. There could be duplicate groups across the responses (by design, not bug).
* The client will be responsible for traversing the responses and building the correct list for the group-selector menu. The API response will include scope details where relevant, but the client will be responsible for determining whether a given group is currently in-scope/allows annotations at the current document URI.

Thanks @robertknight for helping me get to this point!
